Proper hreflang for localized entries in sitemap

edalzell opened the pull request:
### Domain = One Lang

```
<url>
  <loc>https://example.com/about</loc>
</url>
```

### Domain = Multiple Langs & Page = One Lang

#### ie eng only page

```
<url>
  <loc>https://example.com/terms</loc>
  <xhtml:link rel="alternate" hreflang="en" href="https://example.com/terms"/>
</url>
```

#### ie ar only page

```
<url>
  <loc>https://example.com/ar/privacy</loc>
  <xhtml:link rel="alternate" hreflang="ar" href="https://example.com/ar/privacy"/>
</url>
```

### Domain = Multiple Langs & Page = Multiple Langs

```
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml">

  <!-- English (default, no slug) -->
  <url>
    <loc>https://example.com/about</loc>
    <xhtml:link rel="alternate" hreflang="en" href="https://example.com/about"/>
    <xhtml:link rel="alternate" hreflang="ar" href="https://example.com/ar/about"/>
  </url>

  <!-- Arabic -->
  <url>
    <loc>https://example.com/ar/about</loc>
    <xhtml:link rel="alternate" hreflang="en" href="https://example.com/about"/>
    <xhtml:link rel="alternate" hreflang="ar" href="https://example.com/ar/about"/>
  </url>

</urlset>
```

edalzell committed:
Squashed commit of the following:

commit 6f2d84be273a7f2a5c78a076e36a33f732e0c617
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:57:44 2025 -0700

    not used

commit b6ec627b66f0f3d2a96af6c7989d30d90698c0ba
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:57:02 2025 -0700

    not used

commit c695d7c500c78fed369f84f4ddf3e50c2253cfe9
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:49:44 2025 -0700

    fix more tests

commit eaf7b29edc98a2d94cc79205c872bcc150694f6e
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:30:58 2025 -0700

    fix more tests

commit a767b47582dc5020d6b160c3c13cc71dcc038b4d
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:28:46 2025 -0700

    better

commit db53aff8f6e74c89f441b04b28c51bfe077d5ebd
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:13:16 2025 -0700

    start testing

commit 01a30e6a760e237871078718d044df91ef166dae
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:13:10 2025 -0700

    ray

commit a81f79e37b79b767d1f3380ba8dc0587cf44c17c
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:44:10 2025 -0700

    gotta create

commit 8b507ef808f179d9e62a59a58c3c43bd0af3f7cc
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:31:45 2025 -0700

    backwards compatible

commit 15db562be48a28bd7e0d04be8f6df9ab204b0c65
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:31:38 2025 -0700

    tidy

commit 30ab43fe4444d926feeff00e274a8f48839063c6
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:15:32 2025 -0700

    tidy

commit 54bf4df508f920b85abe27e61fbb01cea86b0da3
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:14:20 2025 -0700

    better way

commit 7d7d4e8af2ade39b8a32266309af9a951aa70735
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 12:52:05 2025 -0700

    wip

commit 15ec444b1c8867d3d2ce230ebd38946316a5ba1b
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 11:45:52 2025 -0700

    filter out incorrect domains

edalzell committed:
wip

edalzell committed:
refactor

edalzell committed:
Revert "Squashed commit of the following:"

This reverts commit 5754b09a6dd983812c672e9b117db862e02a57b8.

# Conflicts:
#	src/Sitemap/Sitemap.php
#	tests/Localized/SitemapTest.php

edalzell committed:
add test

edalzell committed:
check for multi site

edalzell committed:
fix tests

edalzell committed:
wip

edalzell committed:
Squashed commit of the following:

commit 6f2d84be273a7f2a5c78a076e36a33f732e0c617
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:57:44 2025 -0700

    not used

commit b6ec627b66f0f3d2a96af6c7989d30d90698c0ba
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:57:02 2025 -0700

    not used

commit c695d7c500c78fed369f84f4ddf3e50c2253cfe9
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:49:44 2025 -0700

    fix more tests

commit eaf7b29edc98a2d94cc79205c872bcc150694f6e
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:30:58 2025 -0700

    fix more tests

commit a767b47582dc5020d6b160c3c13cc71dcc038b4d
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:28:46 2025 -0700

    better

commit db53aff8f6e74c89f441b04b28c51bfe077d5ebd
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:13:16 2025 -0700

    start testing

commit 01a30e6a760e237871078718d044df91ef166dae
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 14:13:10 2025 -0700

    ray

commit a81f79e37b79b767d1f3380ba8dc0587cf44c17c
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:44:10 2025 -0700

    gotta create

commit 8b507ef808f179d9e62a59a58c3c43bd0af3f7cc
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:31:45 2025 -0700

    backwards compatible

commit 15db562be48a28bd7e0d04be8f6df9ab204b0c65
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:31:38 2025 -0700

    tidy

commit 30ab43fe4444d926feeff00e274a8f48839063c6
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:15:32 2025 -0700

    tidy

commit 54bf4df508f920b85abe27e61fbb01cea86b0da3
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 13:14:20 2025 -0700

    better way

commit 7d7d4e8af2ade39b8a32266309af9a951aa70735
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 12:52:05 2025 -0700

    wip

commit 15ec444b1c8867d3d2ce230ebd38946316a5ba1b
Author: edalzell <erin@silentz.co>
Date:   Thu Aug 7 11:45:52 2025 -0700

    filter out incorrect domains

edalzell committed:
fix test

edalzell commented:
Sorry @jesseleite this is ready to go but I'm not sure why the test fails. I'll wait until the other one is merged and give this another look.

edalzell committed:
support x-default in page meta

edalzell committed:
put at end to keep diff smaller

edalzell committed:
fix tests

edalzell committed:
x-default should be on the default site, all the time

edalzell committed:
Merge branch 'master' into fix/localized-entries

# Conflicts:
#	resources/views/meta.antlers.html

edalzell committed:
fix merge

edalzell committed:
fix test

edalzell committed:
Merge branch '6.x' into fix/localized-entries

# Conflicts:
#	src/Http/Controllers/SitemapController.php
#	src/Sitemap/Sitemap.php
#	tests/Fixtures/site-localized/resources/sites.yaml
#	tests/Localized/MetaTagTest.php
#	tests/Localized/SitemapTest.php

edalzell committed:
lint?

edalzell committed:
Fix merge

edalzell committed:
Fix tests

edalzell committed:
add missing namespaces

edalzell committed:
fix multi-site on a diff domain sitemap

edalzell committed:
pass test

edalzell committed:
is it L10?

edalzell committed:
Don’t test L10 on Windows either

duncanmcclean committed:
add back laravel 10 tests

duncanmcclean committed:
avoid escaping with `assertSeeInOrder`

otherwise, the tests fail on laravel 10.

related to this change: laravel/framework#52627

duncanmcclean committed:
wip

duncanmcclean committed:
wip

duncanmcclean committed:
missed a spot

edalzell committed:
test for multiple sites on same domain

duncanmcclean committed:
remove duplicate variable

duncanmcclean committed:
strict

duncanmcclean committed:
use Site facade for getting site

duncanmcclean committed:
combine

duncanmcclean committed:
formatting

duncanmcclean committed:
uncomment these assertions

duncanmcclean committed:
formatting

duncanmcclean committed:
uncomment

duncanmcclean committed:
move hreflangs into existing sitemap template

duncanmcclean committed:
fix xml errors

duncanmcclean committed:
Implement `x-default` link in sitemap

duncanmcclean committed:
update assertions

duncanmcclean committed:
Merge branch 'fix/localized-entries' of github.com:edalzell/seo-pro into fix/localized-entries

duncanmcclean committed:
wip

duncanmcclean committed:
add `x-default` to localized sitemap tests

duncanmcclean committed:
Ensure hreflangs can be generated for taxonomy terms

Before this commit, a type error would be thrown due to the Entry typehint on the `hrefLangs` method.

This commit makes that method more generic so we can support terms too.

This commit also imports the `Entry` contract, rather than the implementation, otherwise it might fail when storing entries in the database.

duncanmcclean committed:
don't need to call `->getContent()` here

edalzell commented on tests/SitemapTest.php:
```diff
@@ -262,95 +245,60 @@ public function it_outputs_paginated_sitemap_page_xml()
             ->assertOk()
             ->assertHeader('Content-Type', 'text/xml; charset=UTF-8')
-            ->getContent();
+            ->assertSeeInOrder([
```
The method used doesn't work when there are `xhtml` tags in the content, so I skipped it due to a lack of time to figure out how to redo. Although thinking about it now, I think `xml_to_array` might work for it.

edalzell commented on src/Sitemap/Sitemap.php:
```diff
@@ -227,4 +235,71 @@ protected function getSiteDefaults()
+    {
+        $sitesWithSameDomain = $this->sitesWithSameDomain($entry->site());
+
+        return $sitesWithSameDomain
```
No, a sitemap should only have data from one domain, see https://github.com/statamic/seo-pro/pull/399

edalzell replied:
The whole reason for these are errors we got when analyzing our clients multi-site SEO. However, I personally, no nothing about SEO and just following @adnankussair 's guidance.

If you are going to include them, please make it so we can take them out with a config/.env var.

adnankussair replied:
## Additional Use Case: Arabic Variant on Site A `example.com`

Imagine adding a new language variant:

```html
<link rel="alternate" hreflang="ar" href="https://example.com/ar/page" />
```

Since `ar` and `en-us` are both hosted on `example.com`, they can share the same sitemap (e.g. `example.com/sitemap.xml`).

Meanwhile, other locales like `en-gb` and `de-de`, hosted on separate domains, will have their own sitemaps.

### Updated `hreflang` block for Site A:

```html
<link rel="alternate" hreflang="en-us" href="https://example.com/page" />
<link rel="alternate" hreflang="en-gb" href="https://example.co.uk/page" />
<link rel="alternate" hreflang="de-de" href="https://example.de/page" />
<link rel="alternate" hreflang="ar" href="https://example.com/ar/page" />
<link rel="alternate" hreflang="x-default" href="https://example.com/page" />
```

> ✅ Same-domain locales can share a sitemap.  
> ✅ Cross-domain locales should use separate sitemaps submitted via their respective Search Console properties.

edalzell committed:
Merge branch 'statamic:6.x' into fix/localized-entries

edalzell committed:
set multi-site hreflangs properly

edalzell committed:
failing test

edalzell committed:
Actually failing test

edalzell committed:
Use new site is default

edalzell committed:
Pass tests

duncanmcclean committed:
wip

duncanmcclean committed:
Remove condition around default hreflang

Now that we're checking against all sites, the condition would always be true, so the condition doesn't need to be there.

duncanmcclean committed:
Revert "Remove condition around default hreflang"

This reverts commit ce61789e67fcceb2c03852528003dc52aa7217b9.

duncanmcclean committed:
Make terms work the same way

duncanmcclean committed:
refactor

duncanmcclean committed:
Add other pages back into this assertion

Before adopting `->assertSeeInOrder()` here, we were asserting that 5 pages were returned.

However, after the method was adopted in this PR, two of the pages were missing from the assertion. This commit adds them back.

duncanmcclean committed:
use short syntax because erin told me to
